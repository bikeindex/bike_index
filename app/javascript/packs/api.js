/*
  Private
*/

const url = urn => [process.env.BASE_URL, urn].join("/");

const serialSearchUrl = serial =>
  url(`api/v2/bikes_search?serial=${serial}`);

const fuzzySearchUrl = serial =>
  url(`api/v2/bikes_search/close_serials?serial=${serial}`);

const partialMatchSerialSearchUrl = params => {
  const query = queryString(params)
  return url(`api/v3/search/serials_containing?${query}`);
}

const serialCloseSearchUrl = params => {
  const query = queryString(params)
  return url(`api/v3/search/close_serials?${query}`);
}

const serialExternalSearchUrl = ({ serial }) => {
  const query = queryString({ serial })
  return url(`api/v3/search/external_registries?${query}`);
}

const request = async url => {
  const resp = await fetch(url);
  const json = await resp.json();
  return json;
};

const queryString = ({ serial, stolenness, location, query, query_items }) => {
  const params = {};
  if (serial) { params.serial = serial; }
  if (stolenness) { params.stolenness = stolenness; }
  if (location) { params.location = location; }
  if (query) { params.query = query; }
  if (query_items) { params.query_items = query_items; }

  return Object.keys(params).map(k => `${k}=${params[k]}`).join("&");
}

/*
  Public
*/

const fetchSerialResults = serial => {
  const url = serialSearchUrl(serial);
  return request(url);
};

const fetchFuzzyResults = serial => {
  const url = fuzzySearchUrl(serial);
  return request(url);
};

const fetchPartialMatchSearch = interpretedParams => {
  const url = partialMatchSerialSearchUrl(interpretedParams);
  return request(url);
}

const fetchSerialCloseSearch = interpretedParams => {
  const url = serialCloseSearchUrl(interpretedParams);
  return request(url);
}

const fetchSerialExternalSearch = ({ raw_serial }) => {
  const url = serialExternalSearchUrl({ serial: raw_serial });
  return request(url);
}

export default {
  fetchSerialResults,
  fetchFuzzyResults,
  fetchSerialExternalSearch,
  fetchSerialCloseSearch,
  fetchPartialMatchSearch,
};
