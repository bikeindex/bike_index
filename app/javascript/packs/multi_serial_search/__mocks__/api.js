const fetchSerialResults = async () => new Promise(resolve => resolve({}));

export {
  fetchSerialResults,
};
