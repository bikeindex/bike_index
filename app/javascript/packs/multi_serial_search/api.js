/*
  Private
*/

const url = urn => [process.env.BASE_URL, urn].join('/');

const serialSearchUrl = serial => url(
  `api/v2/bikes_search?serial=${serial}`,
);

const fuzzySearchUrl = serial => url(
  `api/v2/bikes_search/close_serials?serial=${serial}`,
);

const request = async url => {
  const resp = await fetch(url);
  const json = await resp.json();
  return json;
};

/*
  Public
*/

const fetchSerialResults = serial => {
  const url = serialSearchUrl(serial);
  return request(url);
};

const fetchFuzzyResults = serial => {
  const url = fuzzySearchUrl(serial);
  return request(url);
};

export {
  fetchSerialResults,
  fetchFuzzyResults,
};
