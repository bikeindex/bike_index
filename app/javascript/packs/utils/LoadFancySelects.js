import selectize from "selectize";

// Right now we're using selectize.js (https://selectize.github.io/selectize.js/) to make our select boxes fancy.
// This could be changed to some other library at some point - this makes it possible for us to abstract that away.
// Just add the `unfancy` and `fancy-select` classes to a select box and it will be a fancy select box!
const LoadFancySelects = () => {
  // Gross, check if
  $(".unfancy.fancy-select.no-restore-on-backspace select").selectize({
    create: false,
    plugins: []
  });

  $(".unfancy.fancy-select select").selectize({
    create: false,
    plugins: ["restore_on_backspace"]
  });
  // Remove them so we don't initialize twice
  $(".unfancy.fancy-select, .unfancy.fancy-select-placeholder").removeClass(
    "unfancy"
  );
};

export default LoadFancySelects;
