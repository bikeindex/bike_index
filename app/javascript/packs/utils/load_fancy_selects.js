import selectize from "selectize";

// Right now we're using selectize.js (https://selectize.github.io/selectize.js/) to make our select boxes fancy.
// This could be changed to some other library at some point - this makes it possible for us to abstract that away.
// Just add the `unfancy` and `fancy-select` classes to a select box and it will be a fancy select box!
const LoadFancySelects = () => {
  $(".unfancy.fancy-select.no-restore-on-backspace select").selectize({
    create: false,
    plugins: []
  });
  // Remove, because we've already added them
  $(".unfancy.fancy-select.no-restore-on-backspace").removeClass("unfancy");

  $(".unfancy.fancy-select select").selectize({
    create: false,
    plugins: ["restore_on_backspace"]
  });
  // Remove them so we don't initialize twice
  $(".unfancy.fancy-select").removeClass("unfancy");
};

export default LoadFancySelects;
