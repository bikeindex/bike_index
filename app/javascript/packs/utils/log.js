import loglevel from "loglevel";

if (process.env.RAILS_ENV !== "production") loglevel.setLevel("debug");

export default loglevel;
