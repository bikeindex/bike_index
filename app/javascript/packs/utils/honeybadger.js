import Honeybadger from "honeybadger-js";
import log from "./log";

const environment = process.env.RAILS_ENV;

/* webpacker can't access this exported env */
const apiKey =
  environment === "production" ? process.env.HONEYBADGER_FRONTEND_API_KEY : "foo";

const config = {
  apiKey,
  environment: process.env.RAILS_ENV,
  disabled: environment !== "production"
};
const honeybadger = Honeybadger.configure(config);

Honeybadger.beforeNotify(notice => {
  if (environment === "development") {
    return log.debug("Error", notice);
  }
});
export default honeybadger;
