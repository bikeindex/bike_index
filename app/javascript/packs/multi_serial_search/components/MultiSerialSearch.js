/* eslint import/no-unresolved: 0 */

import React, { useState, Fragment } from "react";
import searchIcon from "images/stolen/search.svg";
import SearchResults from "./SearchResults";
import api from "../../api";
import honeybadger from "../../utils/honeybadger";

const MultiSerialSearch = () => {
  const [serialResults, setSerialResults] = useState(null);
  const [searchTokens, setSearchTokens] = useState("");
  const [loading, setLoading] = useState(false);
  const [fuzzySearching, setFuzzySearching] = useState(false);

  const handleEventErrors = error => {
    honeybadger.notify(error, { component: "MultiSerialSearch" });
    setLoading(false);
  };

  const onSearchSerials = async () => {
    if (!searchTokens) return;
    setLoading(true);

    /*
      normalize the textarea input
    */
    const tokens = searchTokens
      .split(/,|\n/)
      .map(s => s.trim())
      .filter(s => s);
    const uniqSerials = [...new Set(tokens)];

    try {
      /*
        parallel request serials
      */
      const all = await Promise.all(
        uniqSerials.map(serial => api.fetchSerialResults(serial))
      );
      const results = all.map(({ bikes }, index) => {
        const serial = uniqSerials[index];
        return {
          bikes,
          serial,
          fuzzyBikes: [],
          anchor: `#${encodeURI(serial)}`
        };
      });
      setSerialResults(results);
      setFuzzySearching(false);
      setLoading(false);
    } catch (e) {
      handleEventErrors(e);
    }
  };

  const onChangeSearchTokens = e => {
    e.preventDefault();
    const tokens = e.target.value;
    setSearchTokens(tokens);
  };

  const onFuzzySearch = async () => {
    if (fuzzySearching) return;
    if (!serialResults) return;
    setLoading(true);

    try {
      /*
        parallel request fuzzy serials and merge
      */
      const fuzzyAll = await Promise.all(
        serialResults.map(({ serial }) => api.fetchFuzzyResults(serial))
      );
      const updatedResults = fuzzyAll.map(({ bikes: fuzzyBikes }, index) => {
        const serialResult = serialResults[index];
        return Object.assign(serialResult, { fuzzyBikes });
      });
      setSerialResults(updatedResults);
      setFuzzySearching(true);
      setLoading(false);
    } catch (e) {
      handleEventErrors(e);
    }
  };

  return (
    <Fragment>
      <div className="multiserial-form">
        {/* Form  */}
        <span className="padded" />
        <h3>Multiple Serial Search</h3>
        <textarea
          value={searchTokens}
          className="form-control"
          onChange={onChangeSearchTokens}
          placeholder="Enter multiple serial numbers. Separate them with commas or new lines"
        />
        <div className="clearfix">
          <button
            type="submit"
            className="multiserialsearch-btn btn btn-primary float-right mt-2"
            disabled={loading}
            onClick={onSearchSerials}
          >
            <img alt="search" src={searchIcon} />
          </button>
        </div>

        {/* Search Results */}
        <SearchResults
          loading={loading}
          serialResults={serialResults}
          fuzzySearching={fuzzySearching}
          onFuzzySearch={onFuzzySearch}
        />
      </div>
    </Fragment>
  );
};

export default MultiSerialSearch;
