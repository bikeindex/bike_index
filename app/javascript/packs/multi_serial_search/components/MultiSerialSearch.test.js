/* eslint import/no-unresolved: 0 */

import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { bikeResponse } from "helpers/bikes";
import { event } from "helpers/utils";
import api from "../../api";
import MultiSerialSearch from "./MultiSerialSearch";

/*
  api mocks
*/
jest.mock("../../api");
api.fetchSerialResults.mockImplementation(
  () => new Promise(resolve => resolve(bikeResponse))
);

describe("<MultiSerialSearch /> shallow rendered", () => {
  it("matches the snapshot", () => {
    const tree = shallow(<MultiSerialSearch />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it("searches multiple serials", () => {
    jest.spyOn(api, "fetchSerialResults");

    // - toggle form visibility
    const wrap = shallow(<MultiSerialSearch />);

    // - enter and submit multiple serials
    const input = wrap.find("textarea");
    input.simulate("change", event({ target: { value: "one,two,three" } }));
    wrap.find(".multiserial-form button").simulate("click");

    // one call per serial delimited by comma
    expect(api.fetchSerialResults).toHaveBeenCalledTimes(3);
  });
});
