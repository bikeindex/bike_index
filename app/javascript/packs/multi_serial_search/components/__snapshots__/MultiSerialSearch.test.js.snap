// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSerialSearch /> shallow rendered matches the snapshot 1`] = `
<Fragment>
  <h4
    className="multi-search-toggle"
  >
    <a
      href=""
      onClick={[Function]}
    >
      Show
       Multiple Serial Search
    </a>
  </h4>
</Fragment>
`;
